### PR TITLE
Add a Varnish setup for Lando, #355.

### PR DESCRIPTION
**Pull request by tormi**
*Link to ticket: #355*

*Changes proposed in this PR:*
- add a local Varnish service

*How to test:*
1. uncomment the Varnish service definition at `.lando.yml`,
2. rebuild the local environment after adding new service: `lando rebuild -y`,
3. test out `varnishd`: `lando ssh -s varnish -u root -c "varnishd -C -b varnish.drupal-project.lndo.site"`.

### CODE REVIEW

**Comment by tormi** [comment]
`lando ssh -s varnish -c "varnishd -C"` gives currently:

> Error: Cannot create test-file in /var/lib/varnish/a5e95a4f304b/ (Permission denied)
> Check permissions (or delete old directory)

**Comment by tormi** [comment]
Looks like we need to use `root` user here:

```sh
/app$ ls -la /var/lib/varnish
total 12
drwxr-x---    3 varnish  varnish       4096 May  9 09:22 .
drwxr-xr-x    1 root     root          4096 Mar 30 11:45 ..
drwxr-xr-x    3 root     root          4096 May  9 09:22 a5e95a4f304b
```

**Comment by tormi** [comment]
@sampoturve , works with `root`: `lando ssh -s varnish -u root`

```sh
bash-5.1# varnishd -C
Error: -C only good with -b or -f
bash-5.1# varnishd -b
varnishd: option requires an argument: b
usage: varnishd [options]
    -a address[:port][,proto]    # HTTP listen address and port (default: *:80)
                                 #   address: defaults to loopback
                                 #   port: port or service (default: 80)
                                 #   proto: HTTP/1 (default), PROXY
    -b address[:port]            # backend address and port
                                 #   address: hostname or IP
                                 #   port: port or service (default: 80)
    -C                           # print VCL code compiled to C language
    -d                           # debug
    -F                           # Run in foreground
    -f file                      # VCL script
    -h kind[,hashoptions]        # Hash specification
                                 #   -h critbit [default]
                                 #   -h simple_list
                                 #   -h classic
                                 #   -h classic,<buckets>
    -i identity                  # Identity of varnish instance
    -j jail[,jailoptions]        # Jail specification
                                 #   -j unix[,user=<user>][,ccgroup=<group>]
                                 #   -j none
    -l vsl[,vsm]                 # Size of shared memory file
                                 #   vsl: space for VSL records [80m]
                                 #   vsm: space for stats counters [1m]
    -M address:port              # Reverse CLI destination
    -n dir                       # varnishd working directory
    -P file                      # PID file
    -p param=value               # set parameter
    -r param[,param...]          # make parameter read-only
    -S secret-file               # Secret file for CLI authentication
    -s [name=]kind[,options]     # Backend storage specification
                                 #   -s malloc[,<size>]
                                 #   -s file,<dir_or_file>
                                 #   -s file,<dir_or_file>,<size>
                                 #   -s file,<dir_or_file>,<size>,<granularity>
                                 #   -s persistent (experimental)
    -T address:port              # Telnet listen address and port
    -t TTL                       # Default TTL
    -V                           # version
    -W waiter                    # Waiter implementation
                                 #   -W epoll
                                 #   -W poll
    -x optstring                 # List of getopt options
```

**Comment by tormi** [comment]
@sampoturve, `lando ssh -s varnish -u root` & `varnishd -C -b varnish.drupal-project.lndo.site:6082` should work now.

File: .lando.yml [hunk 1]
```diff
@@ -11,6 +11,7 @@ config:
   config:
     php: .lando/php.ini
     database: .lando/my.cnf
+    vcl: .lando/varnish.vcl
```
Comment (sampoturve) [inline .lando.yml:14]: Maybe I'm missing something but with a clean checkout I ran a destroy + start and still can't get the custom headers to appear when the config is here on the top level. When placing into `services → varnish → config → vcl` config loads just fine.

**Missing headers on dev tools**
varnish-missing-vcl-header-lando

File: .lando/varnish.vcl [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+# The routine when we deliver the HTTP request to the user
+# Last chance to modify headers that are sent to the client
+sub vcl_deliver {
+  # Called before a cached object is delivered to the client.
+
+  if (obj.hits > 0) { # Add debug header to see if it's a HIT/MISS and the number of hits, disable when not needed
+    set resp.http.X-W-Cache = "HIT";
+  } else {
+    set resp.http.X-W-Cache = "MISS";
+  }
```
Comment (sampoturve) [inline .lando/varnish.vcl:11]: a) could we add `set resp.http.X-W-Cache-Hits = obj.hits;` for basic debugging too, used in [varnish-configmap-vcl.yaml](https://github.com/wunderio/charts/blob/9528571399354091741263577db92b313afb484e/drupal/templates/varnish-configmap-vcl.yaml#L497) as well.
b) should this have a closing bracket?

Comment (tormi) [reply .lando/varnish.vcl:11]: Thanks for noticing the missing bracket ;)

File: .lando.yml (hunk 1 above)
Comment (tormi) [reply .lando.yml]: Looks like works now. Thanks!
Screenshot 2023-05-11 at 07 28 25

**Review by sampoturve — APPROVED** [review]
Everything works when uncommenting the Varnish configs and running a rebuild. To get a basic site to simply cache content to Varnish, setting a value for "Browser and proxy cache maximum age" in `/admin/config/development/performance` makes Varnish act on it.

Purging is of course another thing and as @misterjoonas mentiones [here](https://github.com/wunderio/drupal-project/pull/347#pullrequestreview-1358384486), the purging features could be added when D10 support is mature enough (looks like it could be now). That could be its own issue.

Basic support is now in and seems to be running well, thank you very much @tormi! 💯

**Comment by tormi** [comment]
Thanks, @sampoturve!

**Pull request merged**